services/redis/queue.js

```diff
@@ -57,15 +59,27 @@ const registerV2ServiceQueue = async (token, userId, follows, topics, locations,
   }
 };
 
-const followTopicServiceQueue = async (user_id, topic_id) => {
+const followTopicServiceQueue = async (user_id, topic_id, delay) => {
   const data = {
     user_id,
     topic_id
   };
 
+  let currentTime = momentTz().tz('America/Los_Angeles');
+  const randomTime = sample([6, 7, 8]);
+  const additionalDays = delay;
+
+  let requiredTime = momentTz()
+    .tz('America/Los_Angeles')
+    .set({hour: randomTime})
+    .add(additionalDays, 'days');
+
+  const diffTime = requiredTime.diff(currentTime, 'milliseconds');
+
   const options = {
     jobId: uuidv4(),
-    removeOnComplete: true
+    removeOnComplete: true,
+    delay: diffTime
```

Comment:
<img src="https://avatars.githubusercontent.com/in/347564?s=41" alt="Image description" width="20" height="20">   CodeRabbit

The logic for calculating the delay time seems to be correct, but there are a few potential issues:

1. The `randomTime` variable is set to a random value from the array `[6, 7, 8]`. This means that the hour of execution will always be either 6, 7, or 8 AM. If this is the intended behavior, then it's fine. However, if you want a truly random hour of execution, you might want to consider using a different method to generate this value.

2. The `currentTime` and `requiredTime` variables are both set to the current time in the 'America/Los_Angeles' timezone. This could potentially cause issues if your users are located in different timezones. You might want to consider adjusting these times based on the user's location.

3. The `diffTime` variable is calculated as the difference between `requiredTime` and `currentTime` in milliseconds. This is used as the delay for the queue. However, if `requiredTime` is less than `currentTime`, then `diffTime` will be negative, which could cause unexpected behavior. You should add a check to ensure that `diffTime` is never less than zero.

Here's a possible fix:

```diff
- let currentTime = momentTz().tz('America/Los_Angeles');
+ let currentTime = momentTz().tz(user.timezone);
  
- const randomTime = sample([6, 7, 8]);
+ const randomTime = Math.floor(Math.random() * 24); // Random hour of the day
  
- let requiredTime = momentTz()
-   .tz('America/Los_Angeles')
+ let requiredTime = momentTz()
+   .tz(user.timezone)
    .set({hour: randomTime})
    .add(additionalDays, 'days');
  
- const diffTime = requiredTime.diff(currentTime, 'milliseconds');
+ let diffTime = requiredTime.diff(currentTime, 'milliseconds');
+ if (diffTime < 0) {
+   diffTime = 0;
+ }
```

Please note that you would need to fetch the user's timezone (`user.timezone`) before using it in this code.


<!-- This is an auto-generated comment by OSS CodeRabbit -->